fe/spark-dpp/src/main/java/com/starrocks/load/loadv2/dpp/SparkDpp.java

```diff
@@ -610,48 +610,90 @@ public boolean validateData(Object srcValue, EtlJobConfig.EtlColumn etlColumn, C
             }
         }
+        // check fields name and size
+        HashSet<String> schemaSet = new HashSet<>(Arrays.asList(sourceData.schema().fieldNames()));
+        schemaSet.addAll(Arrays.asList(srcSchema.fieldNames()));
+        if (sourceData.schema().size() != schemaSet.size()) {
+            throw new SparkDppException("The schema of file and load statement must be equal. " +
+                    "file schema: " + srcSchema.treeString() + ", table schema: " + sourceData.schema().treeString());
+        }
+        scannedRowsAcc.add(sourceData.count());
+        // TODO: data quality check for orc/parquet load
+        // Check process is roughly the same as the hive load, but there are some bugs to fix.
+        // Uncomment below when method checkDataFromHiveWithStrictMode is ready.
+        // checkDataFromHiveWithStrictMode(sourceData, baseIndex, fileGroup.columnMappings.keySet(), etlJobConfig.properties.strictMode,
+        //        dstTableSchema, dictBitmapColumnSet);
+        return sourceData;
+    }
+
+    /**
+     * Note: parameter fileUrl cannot contain any wildcards to prevent output of wrong results.
+     * For example, if fileUrl is specified by regular expression like hdfs://some/p1=*",
+     * column value will be extracted as "*" in {@link DppUtils#parseColumnsFromPath},
+     * then "*" will likely be cast to wrong output according to its field type such as NULL for INT
+     * in {@link SparkDpp#convertSrcDataframeToDstDataframe}
+     */
+    private Dataset<Row> loadDataFromPath(SparkSession spark,
+                                          EtlJobConfig.EtlFileGroup fileGroup,
+                                          String fileUrl,
+                                          EtlJobConfig.EtlIndex baseIndex,
+                                          List<EtlJobConfig.EtlColumn> columns) throws SparkDppException {
+        List<String> columnValueFromPath = DppUtils.parseColumnsFromPath(fileUrl, fileGroup.columnsFromPath);
         // for getting schema to check source data
         Map<String, Integer> dstColumnNameToIndex = new HashMap<String, Integer>();
         for (int i = 0; i < baseIndex.columns.size(); i++) {
             dstColumnNameToIndex.put(baseIndex.columns.get(i).columnName, i);
         }
-        List<String> srcColumnsWithColumnsFromPath = new ArrayList<>();
-        srcColumnsWithColumnsFromPath.addAll(dataSrcColumns);
-        if (fileGroup.columnsFromPath != null) {
-            srcColumnsWithColumnsFromPath.addAll(fileGroup.columnsFromPath);
-        }
-        StructType srcSchema = createScrSchema(srcColumnsWithColumnsFromPath);
+        StructType srcSchema = constructSrcSchema(fileGroup, baseIndex);
         JavaRDD<String> sourceDataRdd = spark.read().textFile(fileUrl).toJavaRDD();
-        int columnSize = dataSrcColumns.size();
+        int columnSize = fileGroup.fileFieldNames.size();
```

Comment:
You cannot directly use fileGroup.fileFieldNames, because it may be null.

Reply:
fixed, PTAL